rust/lance-encoding/src/encodings/logical/primitive.rs

```diff
@@ -242,6 +246,727 @@ impl FieldScheduler for PrimitiveFieldScheduler {
+#[derive(Debug)]
+struct MiniBlockDecoder {
+    rep_decompressor: Arc<dyn BlockDecompressor>,
+    def_decompressor: Arc<dyn BlockDecompressor>,
+    value_decompressor: Arc<dyn MiniBlockDecompressor>,
+    data: VecDeque<ScheduledChunk>,
+    offset_in_current_chunk: u64,
+    num_rows: u64,
+}
+
+impl StructuralPageDecoder for MiniBlockDecoder {
+    fn drain(&mut self, num_rows: u64) -> Result<Box<dyn DecodePageTask>> {
+        let mut remaining = num_rows;
+        let mut chunks = Vec::new();
+        let offset_into_first_chunk = self.offset_in_current_chunk;
+        while remaining > 0 {
+            if remaining >= self.data.front().unwrap().vals_in_chunk - self.offset_in_current_chunk
+            {
+                // We are fully consuming the next chunk
+                let chunk = self.data.pop_front().unwrap();
+                remaining -= chunk.vals_in_chunk - self.offset_in_current_chunk;
+                chunks.push(chunk);
+                self.offset_in_current_chunk = 0;
+            } else {
+                // We are partially consuming the next chunk
+                let chunk = self.data.front().unwrap().clone();
+                self.offset_in_current_chunk += remaining;
+                debug_assert!(self.offset_in_current_chunk > 0);
+                remaining = 0;
+                chunks.push(chunk);
+            }
+        }
+        Ok(Box::new(DecodeMiniBlockTask {
+            chunks,
+            rep_decompressor: self.rep_decompressor.clone(),
+            def_decompressor: self.def_decompressor.clone(),
+            value_decompressor: self.value_decompressor.clone(),
+            num_rows,
+            offset_into_first_chunk,
+        }))
+    }
+
+    fn num_rows(&self) -> u64 {
+        self.num_rows
+    }
+}
+
+/// A scheduler for all-null data
+///
+/// Note that all-null data might still require buffers.  If there are definition levels
+/// then we need to distinguish between null structs and null values.  If there are repetition
+/// levels then we need to distinguish between null lists, lists of null, and empty lists.
+#[derive(Debug, Default)]
+pub struct AllNullScheduler {}
+
+impl StructuralPageScheduler for AllNullScheduler {
+    fn initialize<'a>(&'a mut self, _io: &Arc<dyn EncodingsIo>) -> BoxFuture<'a, Result<()>> {
+        std::future::ready(Ok(())).boxed()
+    }
+
+    fn schedule_ranges(
+        &self,
+        _ranges: &[Range<u64>],
+        _io: &dyn EncodingsIo,
+    ) -> Result<BoxFuture<'static, Result<Box<dyn StructuralPageDecoder>>>> {
+        Ok(std::future::ready(Ok(
+            Box::new(AllNullPageDecoder {}) as Box<dyn StructuralPageDecoder>
+        ))
+        .boxed())
+    }
+}
+
+#[derive(Debug)]
+struct AllNullDecodePageTask {}
+impl DecodePageTask for AllNullDecodePageTask {
+    fn decode(self: Box<Self>) -> Result<DecodedPage> {
+        // TODO: Not that trivial, we might have rep/def that we need to encode / decode still
+        todo!()
```

Comment:
Should these be filled in for this PR? Or are we just not writing them yet so it doesn't matter?

Reply:
Yeah, there are a fair number of TODO's in this PR that I'm leaving for future PRs.